main.cpp

```diff
@@ -1,84 +1,16 @@
-        "Blue", "Orange", "Green", "Brown", "Slate"
-    };
-    int numberOfMinorColors =
-        sizeof(MinorColorNames) / sizeof(MinorColorNames[0]);
-
-    class ColorPair {
-        private:
-            MajorColor majorColor;
-            MinorColor minorColor;
-        public:
-            ColorPair(MajorColor major, MinorColor minor):
-                majorColor(major), minorColor(minor)
-            {}
-            MajorColor getMajor() {
-                return majorColor;
-            }
-            MinorColor getMinor() {
-                return minorColor;
-            }
-            std::string ToString() {
-                std::string colorPairStr = MajorColorNames[majorColor];
-                colorPairStr += " ";
-                colorPairStr += MinorColorNames[minorColor];
-                return colorPairStr;
-            }
-    };
-
-    ColorPair GetColorFromPairNumber(int pairNumber) {
-        int zeroBasedPairNumber = pairNumber - 1;
-        MajorColor majorColor = 
-            (MajorColor)(zeroBasedPairNumber / numberOfMinorColors);
-        MinorColor minorColor =
-            (MinorColor)(zeroBasedPairNumber % numberOfMinorColors);
-        return ColorPair(majorColor, minorColor);
-    }
-    int GetPairNumberFromColor(MajorColor major, MinorColor minor) {
-        return major * numberOfMinorColors + minor + 1;
-    }
-}
-
-void testNumberToPair(int pairNumber,
-    TelCoColorCoder::MajorColor expectedMajor,
-    TelCoColorCoder::MinorColor expectedMinor)
-{
-    TelCoColorCoder::ColorPair colorPair =
-        TelCoColorCoder::GetColorFromPairNumber(pairNumber);
-    std::cout << "Got pair " << colorPair.ToString() << std::endl;
-    assert(colorPair.getMajor() == expectedMajor);
-    assert(colorPair.getMinor() == expectedMinor);
-}
-
-void testPairToNumber(
-    TelCoColorCoder::MajorColor major,
-    TelCoColorCoder::MinorColor minor,
-    int expectedPairNumber)
-{
-    int pairNumber = TelCoColorCoder::GetPairNumberFromColor(major, minor);
-    std::cout << "Got pair number " << pairNumber << std::endl;
-    assert(pairNumber == expectedPairNumber);
-}
-
-int main() {
-    testNumberToPair(4, TelCoColorCoder::WHITE, TelCoColorCoder::BROWN);
-    testNumberToPair(5, TelCoColorCoder::WHITE, TelCoColorCoder::SLATE);
-
-    testPairToNumber(TelCoColorCoder::BLACK, TelCoColorCoder::ORANGE, 12);
-    testPairToNumber(TelCoColorCoder::VIOLET, TelCoColorCoder::SLATE, 25);
-
-    return 0;
-}
+#include "TestColorCoding.h"
+
+int main() {
+    using namespace TelCoColorCoder;
+
+    TestColorCoding::testNumberToPair(4, MajorColor::WHITE, MinorColor::BROWN);
+    TestColorCoding::testNumberToPair(5, MajorColor::WHITE, MinorColor::SLATE);
+
+    TestColorCoding::testPairToNumber(MajorColor::BLACK, MinorColor::ORANGE, 12);
+    TestColorCoding::testPairToNumber(MajorColor::VIOLET, MinorColor::SLATE, 25);
+
+    std::cout << "Printing all color codes:" << std::endl;
+    ColorUtils::printColorCoding();

```

Comment:
The `test...` functions above have asserts, which means we don't need to manually see the console to check their functionality.
However, testing of `printColorCoding` requires manual inspection of 25 lines of output. Can you think of some asserts to reduce the manual testing burden?

Reply:
I have removed the print function instead added a test which automatically test the color coding.

Reply:
Yes, I saw the iteration through all the pair numbers. But why remove the print function? You had written it because it was needed, right?
Perhaps it is hard to test that print function? When tests are hard to write, usually it indicates that the code is doing too many diverse things. Time to refactor the code. We will continue this in the next assignment

Reply:
As you mentioned in next assignment have refactored the code.